File: appcenter/android/src/main/java/com/getcapacitor/plugin/appcenter/AppCenterPlugin.kt
package com.getcapacitor.plugin.appcenter

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.microsoft.appcenter.reactnative.shared.AppCenterReactNativeShared

@CapacitorPlugin(name = "AppCenter")
public class AppCenterPlugin : Plugin() {
    private val implementation = AppCenterBase()

    override fun load() {
        AppCenterReactNativeShared.configureAppCenter(activity.application)
    }

    @PluginMethod
    public fun getInstallId(call: PluginCall) {
        val ret = JSObject()
        ret.put("value", implementation.installId)
        call.resolve(ret)
    }

    @PluginMethod(returnType = PluginMethod.RETURN_NONE)
    public fun setUserId(call: PluginCall) {
        implementation.setUserId(call.getString("userId", null))
        call.resolve()
    }

    @PluginMethod
    public fun getSdkVersion(call: PluginCall) {
        val ret = JSObject()
        ret.put("value", implementation.sdkVersion)
        call.resolve(ret)
    }

    @PluginMethod
    public fun isEnabled(call: PluginCall) {
        val ret = JSObject()
        ret.put("value", implementation.isEnabled)
        call.resolve(ret)
    }

    @PluginMethod(returnType = PluginMethod.RETURN_NONE)
    public fun setEnabled(call: PluginCall) {
        implementation.enable(call.getBoolean("enabled", false) ?: false)
        call.resolve()
    }

    @PluginMethod
    public fun getLogLevel(call: PluginCall) {
        val ret = JSObject()
        ret.put("value", implementation.logLevel)
        call.resolve(ret)
    }

    @PluginMethod(returnType = PluginMethod.RETURN_NONE)
    public fun setLogLevel(call: PluginCall) {
        if (!call.data.has("logLevel")) {
            call.reject("Must provide a LogLevel")
            return
        }

        // A logLevel that is not an integer is null here. The Java code threw a NullPointerException
        // while unboxing it, which the `!!` keeps.
        implementation.logLevel = call.getInt("logLevel")!!
        call.resolve()
    }

    @PluginMethod(returnType = PluginMethod.RETURN_NONE)
    public fun setCustomProperties(call: PluginCall) {
        call.unavailable("Not available in appcenter 2.0.0 or later.")
    }

    @PluginMethod(returnType = PluginMethod.RETURN_NONE)
    public fun setNetworkRequestsAllowed(call: PluginCall) {
        implementation.isNetworkRequestsAllowed = call.getBoolean("isAllowed", true) ?: true
        call.resolve()
    }

    @PluginMethod
    public fun isNetworkRequestsAllowed(call: PluginCall) {
        val ret = JSObject()
        ret.put("value", implementation.isNetworkRequestsAllowed)
        call.resolve(ret)
    }

    @PluginMethod(returnType = PluginMethod.RETURN_NONE)
    public fun setCountryCode(call: PluginCall) {
        implementation.setCountryCode(call.getString("countryCode", null))
        call.resolve()
    }
}

File: appcenter/ios/Sources/AppCenterPlugin/AppCenterPlugin.swift
import Foundation
import Capacitor
import AppCenterCapacitorShared

@objc(AppCenterPlugin)
public class AppCenterPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "AppCenterPlugin"
    public let jsName = "AppCenter"
    public let pluginMethods: [CAPPluginMethod] = [
        .promise("getInstallId", AppCenterPlugin.getInstallId),
        .none("setUserId", AppCenterPlugin.setUserId),
        .promise("getSdkVersion", AppCenterPlugin.getSdkVersion),
        .promise("isEnabled", AppCenterPlugin.isEnabled),
        .none("setEnabled", AppCenterPlugin.setEnabled),
        .none("setCustomProperties", AppCenterPlugin.setCustomProperties),
        .promise("getLogLevel", AppCenterPlugin.getLogLevel),
        .none("setLogLevel", AppCenterPlugin.setLogLevel),
        .none("setNetworkRequestsAllowed", AppCenterPlugin.setNetworkRequestsAllowed),
        .promise("isNetworkRequestsAllowed", AppCenterPlugin.isNetworkRequestsAllowed),
        .none("setCountryCode", AppCenterPlugin.setCountryCode)
    ]

    private let implementation = AppCenterBase()

    override public func load() {
        AppCenterCapacitorShared.configureWithSettings()
    }

    func getInstallId(_ call: CAPPluginCall) {
        call.resolve(["value": implementation.getInstallId()])
    }

    func setUserId(_ call: CAPPluginCall) {
        implementation.setUserId(call.getString("userId") ?? "")
        call.resolve()
    }

    func getSdkVersion(_ call: CAPPluginCall) {
        call.resolve(["value": implementation.getSdkVersion()])
    }

    func isEnabled(_ call: CAPPluginCall) {
        call.resolve(["value": implementation.isEnabled()])
    }

    func setEnabled(_ call: CAPPluginCall) {
        implementation.enable(call.getBool("enabled") ?? false)
        call.resolve()
    }

    func getLogLevel(_ call: CAPPluginCall) {
        call.resolve(["value": implementation.getLogLevel()])
    }

    func setLogLevel(_ call: CAPPluginCall) throws {
        guard let level = call.options["logLevel"] as? Int else {
            throw CAPPluginError("Must provide LogLevel")
        }

        implementation.setLogLevel(level)
        call.resolve()
    }

    func setCustomProperties(_ call: CAPPluginCall) {
        call.unavailable("Not available in appcenter@2.0.0 or later.")
        //        guard let properties = call.options["properties"] as? [String: [String: Any]] else {
        //            implementation.setCustomProperties([:])
        //            call.resolve()
        //            return
        //        }
        //
        //        implementation.setCustomProperties(properties)
        //
        //        call.resolve()
    }

    func setNetworkRequestsAllowed(_ call: CAPPluginCall) {
        implementation.setNetworkRequestsAllowed(call.getBool("isAllowed", true))
        call.resolve()
    }

    func isNetworkRequestsAllowed(_ call: CAPPluginCall) {
        call.resolve(["value": implementation.isNetWorkRequestsAllowed()])
    }

    func setCountryCode(_ call: CAPPluginCall) {
        implementation.setCountryCode(call.getString("countryCode") ?? "")
        call.resolve()
    }
}
